spark/src/main/scala/org/apache/spark/sql/delta/serverSidePlanning/ServerSidePlanningClient.scala

```diff
@@ -0,0 +1,113 @@
+ */
+case class ScanFile(
+  filePath: String,
+  fileSizeInBytes: Long,
+  fileFormat: String  // "parquet", "orc", etc.
+)
+
+/**
+ * Result of a table scan plan operation.
+ */
+case class ScanPlan(
+  files: Seq[ScanFile]
+)
+
+/**
+ * Interface for planning table scans via server-side planning (e.g., Iceberg REST catalog).
+ * This interface is intentionally simple and has no dependencies
+ * on Iceberg libraries, allowing it to live in delta-spark module.
+ */
+trait ServerSidePlanningClient {
+  /**
+   * Plan a table scan and return the list of files to read.
+   *
+   * @param database The database or schema name
+   * @param table The table name
+   * @return ScanPlan containing files to read
+   */
+  def planScan(database: String, table: String): ScanPlan
+}
+
+/**
+ * Factory for creating ServerSidePlanningClient instances.
+ * This allows for configurable implementations (REST, mock, Spark-based, etc.)
+ */
+trait ServerSidePlanningClientFactory {
+  /**
+   * Create a client for a specific catalog by reading catalog-specific configuration.
+   * This method reads configuration from spark.sql.catalog.<catalogName>.uri and
+   * spark.sql.catalog.<catalogName>.token.
+   *
+   * @param spark The SparkSession
+   * @param catalogName The name of the catalog (e.g., "spark_catalog", "unity")
+   * @return A ServerSidePlanningClient configured for the specified catalog
+   */
+  def buildForCatalog(spark: SparkSession, catalogName: String): ServerSidePlanningClient
+}
+
+/**
+ * Registry for client factories. Can be configured for testing or to provide
+ * production implementations (e.g., IcebergRESTCatalogPlanningClientFactory).
+ *
+ * By default, no factory is registered. Production code should register an appropriate
+ * factory implementation before attempting to create clients.
+ */
+object ServerSidePlanningClientFactory {
+  @volatile private var registeredFactory: Option[ServerSidePlanningClientFactory] = None
+
+  /**
+   * Set a factory for production use or testing.
+   */
+  def setFactory(factory: ServerSidePlanningClientFactory): Unit = {
+    registeredFactory = Some(factory)
+  }
+
+  /**
+   * Clear the registered factory.
+   */
+  def clearFactory(): Unit = {
+    registeredFactory = None
+  }
+
+  /**
+   * Get the currently registered factory.
+   * Throws IllegalStateException if no factory has been registered.
+   */
+  def getFactory(): ServerSidePlanningClientFactory = {
+    registeredFactory.getOrElse {
+      throw new IllegalStateException(
+        "No ServerSidePlanningClientFactory has been registered. " +
+        "Call ServerSidePlanningClientFactory.setFactory() to register an implementation.")
+    }
+  }
+
+  /**
+   * Convenience method to create a client for a specific catalog using the registered factory.
+   */
+  def buildForCatalog(spark: SparkSession, catalogName: String): ServerSidePlanningClient = {
+    getFactory().buildForCatalog(spark, catalogName)
+  }
```

Comment:
lets remove keep only one of getFactory or buildForCatalog. i think a simple getClient should be good.